// @pgsql/quotes/kwlist.js
"use strict";
/* eslint-disable */
/**
 * Generated from PostgreSQL kwlist.h
 * DO NOT EDIT BY HAND.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.TYPE_FUNC_NAME_KEYWORDS = exports.COL_NAME_KEYWORDS = exports.UNRESERVED_KEYWORDS = exports.RESERVED_KEYWORDS = exports.kwlist = void 0;
exports.keywordKindOf = keywordKindOf;
exports.kwlist = {
    UNRESERVED_KEYWORD: [
        "abort",
        "absent",
        "absolute",
        "access",
        "action",
        "add",
        "admin",
        "after",
        "aggregate",
        "also",
        "alter",
        "always",
        "asensitive",
        "assertion",
        "assignment",
        "at",
        "atomic",
        "attach",
        "attribute",
        "backward",
        "before",
        "begin",
        "breadth",
        "by",
        "cache",
        "call",
        "called",
        "cascade",
        "cascaded",
        "catalog",
        "chain",
        "characteristics",
        "checkpoint",
        "class",
        "close",
        "cluster",
        "columns",
        "comment",
        "comments",
        "commit",
        "committed",
        "compression",
        "conditional",
        "configuration",
        "conflict",
        "connection",
        "constraints",
        "content",
        "continue",
        "conversion",
        "copy",
        "cost",
        "csv",
        "cube",
        "current",
        "cursor",
        "cycle",
        "data",
        "database",
        "day",
        "deallocate",
        "declare",
        "defaults",
        "deferred",
        "definer",
        "delete",
        "delimiter",
        "delimiters",
        "depends",
        "depth",
        "detach",
        "dictionary",
        "disable",
        "discard",
        "document",
        "domain",
        "double",
        "drop",
        "each",
        "empty",
        "enable",
        "encoding",
        "encrypted",
        "enforced",
        "enum",
        "error",
        "escape",
        "event",
        "exclude",
        "excluding",
        "exclusive",
        "execute",
        "explain",
        "expression",
        "extension",
        "external",
        "family",
        "filter",
        "finalize",
        "first",
        "following",
        "force",
        "format",
        "forward",
        "function",
        "functions",
        "generated",
        "global",
        "granted",
        "groups",
        "handler",
        "header",
        "hold",
        "hour",
        "identity",
        "if",
        "ignore",
        "immediate",
        "immutable",
        "implicit",
        "import",
        "include",
        "including",
        "increment",
        "indent",
        "index",
        "indexes",
        "inherit",
        "inherits",
        "inline",
        "input",
        "insensitive",
        "insert",
        "instead",
        "invoker",
        "isolation",
        "keep",
        "key",
        "keys",
        "label",
        "language",
        "large",
        "last",
        "leakproof",
        "level",
        "listen",
        "load",
        "local",
        "location",
        "lock",
        "locked",
        "logged",
        "lsn",
        "mapping",
        "match",
        "matched",
        "materialized",
        "maxvalue",
        "merge",
        "method",
        "minute",
        "minvalue",
        "mode",
        "month",
        "move",
        "name",
        "names",
        "nested",
        "new",
        "next",
        "nfc",
        "nfd",
        "nfkc",
        "nfkd",
        "no",
        "normalized",
        "nothing",
        "notify",
        "nowait",
        "nulls",
        "object",
        "objects",
        "of",
        "off",
        "oids",
        "old",
        "omit",
        "operator",
        "option",
        "options",
        "ordinality",
        "others",
        "over",
        "overriding",
        "owned",
        "owner",
        "parallel",
        "parameter",
        "parser",
        "partial",
        "partition",
        "partitions",
        "passing",
        "password",
        "path",
        "period",
        "plan",
        "plans",
        "policy",
        "preceding",
        "prepare",
        "prepared",
        "preserve",
        "prior",
        "privileges",
        "procedural",
        "procedure",
        "procedures",
        "program",
        "publication",
        "quote",
        "quotes",
        "range",
        "read",
        "reassign",
        "recursive",
        "ref",
        "referencing",
        "refresh",
        "reindex",
        "relative",
        "release",
        "rename",
        "repeatable",
        "replace",
        "replica",
        "reset",
        "respect",
        "restart",
        "restrict",
        "return",
        "returns",
        "revoke",
        "role",
        "rollback",
        "rollup",
        "routine",
        "routines",
        "rows",
        "rule",
        "savepoint",
        "scalar",
        "schema",
        "schemas",
        "scroll",
        "search",
        "second",
        "security",
        "sequence",
        "sequences",
        "serializable",
        "server",
        "session",
        "set",
        "sets",
        "share",
        "show",
        "simple",
        "skip",
        "snapshot",
        "source",
        "split",
        "sql",
        "stable",
        "standalone",
        "start",
        "statement",
        "statistics",
        "stdin",
        "stdout",
        "storage",
        "stored",
        "strict",
        "string",
        "strip",
        "subscription",
        "support",
        "sysid",
        "system",
        "tables",
        "tablespace",
        "target",
        "temp",
        "template",
        "temporary",
        "text",
        "ties",
        "transaction",
        "transform",
        "trigger",
        "truncate",
        "trusted",
        "type",
        "types",
        "uescape",
        "unbounded",
        "uncommitted",
        "unconditional",
        "unencrypted",
        "unknown",
        "unlisten",
        "unlogged",
        "until",
        "update",
        "vacuum",
        "valid",
        "validate",
        "validator",
        "value",
        "varying",
        "version",
        "view",
        "views",
        "virtual",
        "volatile",
        "wait",
        "whitespace",
        "within",
        "without",
        "work",
        "wrapper",
        "write",
        "xml",
        "year",
        "yes",
        "zone"
    ],
    RESERVED_KEYWORD: [
        "all",
        "analyse",
        "analyze",
        "and",
        "any",
        "array",
        "as",
        "asc",
        "asymmetric",
        "both",
        "case",
        "cast",
        "check",
        "collate",
        "column",
        "constraint",
        "create",
        "current_catalog",
        "current_date",
        "current_role",
        "current_time",
        "current_timestamp",
        "current_user",
        "default",
        "deferrable",
        "desc",
        "distinct",
        "do",
        "else",
        "end",
        "except",
        "false",
        "fetch",
        "for",
        "foreign",
        "from",
        "grant",
        "group",
        "having",
        "in",
        "initially",
        "intersect",
        "into",
        "lateral",
        "leading",
        "limit",
        "localtime",
        "localtimestamp",
        "not",
        "null",
        "offset",
        "on",
        "only",
        "or",
        "order",
        "placing",
        "primary",
        "references",
        "returning",
        "select",
        "session_user",
        "some",
        "symmetric",
        "system_user",
        "table",
        "then",
        "to",
        "trailing",
        "true",
        "union",
        "unique",
        "user",
        "using",
        "variadic",
        "when",
        "where",
        "window",
        "with"
    ],
    TYPE_FUNC_NAME_KEYWORD: [
        "authorization",
        "binary",
        "collation",
        "concurrently",
        "cross",
        "current_schema",
        "freeze",
        "full",
        "ilike",
        "inner",
        "is",
        "isnull",
        "join",
        "left",
        "like",
        "natural",
        "notnull",
        "outer",
        "overlaps",
        "right",
        "similar",
        "tablesample",
        "verbose"
    ],
    COL_NAME_KEYWORD: [
        "between",
        "bigint",
        "bit",
        "boolean",
        "char",
        "character",
        "coalesce",
        "dec",
        "decimal",
        "exists",
        "extract",
        "float",
        "greatest",
        "grouping",
        "inout",
        "int",
        "integer",
        "interval",
        "json",
        "json_array",
        "json_arrayagg",
        "json_exists",
        "json_object",
        "json_objectagg",
        "json_query",
        "json_scalar",
        "json_serialize",
        "json_table",
        "json_value",
        "least",
        "merge_action",
        "national",
        "nchar",
        "none",
        "normalize",
        "nullif",
        "numeric",
        "out",
        "overlay",
        "position",
        "precision",
        "real",
        "row",
        "setof",
        "smallint",
        "substring",
        "time",
        "timestamp",
        "treat",
        "trim",
        "values",
        "varchar",
        "xmlattributes",
        "xmlconcat",
        "xmlelement",
        "xmlexists",
        "xmlforest",
        "xmlnamespaces",
        "xmlparse",
        "xmlpi",
        "xmlroot",
        "xmlserialize",
        "xmltable"
    ]
};
exports.RESERVED_KEYWORDS = new Set(exports.kwlist.RESERVED_KEYWORD ?? []);
exports.UNRESERVED_KEYWORDS = new Set(exports.kwlist.UNRESERVED_KEYWORD ?? []);
exports.COL_NAME_KEYWORDS = new Set(exports.kwlist.COL_NAME_KEYWORD ?? []);
exports.TYPE_FUNC_NAME_KEYWORDS = new Set(exports.kwlist.TYPE_FUNC_NAME_KEYWORD ?? []);
function keywordKindOf(word) {
    const w = word.toLowerCase();
    if (exports.RESERVED_KEYWORDS.has(w))
        return "RESERVED_KEYWORD";
    if (exports.TYPE_FUNC_NAME_KEYWORDS.has(w))
        return "TYPE_FUNC_NAME_KEYWORD";
    if (exports.COL_NAME_KEYWORDS.has(w))
        return "COL_NAME_KEYWORD";
    if (exports.UNRESERVED_KEYWORDS.has(w))
        return "UNRESERVED_KEYWORD";
    return "NO_KEYWORD";
}
